merkle/compact_merkle_tree.go

```diff
@@ -158,39 +160,45 @@ func (c *CompactMerkleTree) recalculateRoot(f setNodeFunc) {
 				first = false
 			} else {
 				newRoot = c.hasher.HashChildren(c.nodes[bit], newRoot)
-				f(bit+1, index, newRoot)
+				if err := f(bit+1, index, newRoot); err != nil {
+					return err
+				}
 			}
 		}
 		mask <<= 1
 	}
 	c.root = newRoot
+	return nil
 }
 
 // AddLeaf calculates the leafhash of |data| and appends it to the tree.
 // |f| is a callback which will be called multiple times with the full MerkleTree coordinates of nodes whose hash should be updated.
-func (c *CompactMerkleTree) AddLeaf(data []byte, f setNodeFunc) (int64, []byte) {
+func (c *CompactMerkleTree) AddLeaf(data []byte, f setNodeFunc) (int64, []byte, error) {
 	h := c.hasher.HashLeaf(data)
-	return c.AddLeafHash(h, f), h
+	seq, err := c.AddLeafHash(h, f)
```

Comment:
nit: Return "default" values for seq and h if err != nil

if err != nil {
  return 0, nil, err
}